File: Javascript/Searching-Sorting/binarySearch.js
//Divide & Conquerer Technique
//find the index of given no in a sorted array
let a = [11, 12, 13, 14, 15, 16, 17, 18, 19, 20];

const binarySearch = (arr, search, left, right) => {
    while (left <= right) {
        console.log("how many times")
        let mid = Number.parseInt((left + right) / 2);
        if (arr[mid] === search) {
            return mid + 1;
        }
        else if (arr[mid] < search) {
            console.log("left side ")
            left = mid + 1;
        }
        else {
            console.log("right side")
            right = mid - 1;
        }
    }
    return -1;
}

console.log(binarySearch(a, 17, 0, a.length - 1));

File: binary_search.py
def binary_search(arr, target):
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2  # Calculate the middle index

        if arr[mid] == target:
            return mid  # Element found, return its index
        elif arr[mid] < target:
            left = mid + 1  # Adjust the left boundary
        else:
            right = mid - 1  # Adjust the right boundary

    return -1  # Element not found in the array

# Example usage
if __name__ == "__main__":
    # Sorted array
    arr = [1, 3, 5, 7, 9, 11, 13, 15, 17]

    # Target element to search for
    target = 7

    # Perform binary search
    result = binary_search(arr, target)

    if result != -1:
        print(f"Element {target} found at index {result}")
    else:
        print(f"Element {target} not found in the array")
